### PR TITLE
Add sample_weight to eval_metric

### PR DESCRIPTION
**Pull request by BenEfrati**
Hi,
After https://github.com/dmlc/xgboost/pull/6751 provide a custom eval_metric via fit is deprecated.

In our eval_metric function we call to `get_weight()` from dmatrix.
When passing the metric function via constructor or `set_params` method we're unable to get sample_weight.

It might be useful for other sklearn metric function that accept sample_weight argument

### CODE REVIEW

**Comment by idanmoradarthas** [comment]
This is an important ability!

**Review by trivialfis — COMMENTED** [review]
Thank you for the new feature! Let's not add an additional parameter and pass the weight when present.

**Comment by BenEfrati** [comment]
> Thank you for the new feature! Let's not add an additional parameter and pass the weight when present.

Agree,
Are you ok with the try except approach?

**Comment by trivialfis** [comment]
I think we can patch the `_metric_decorator` as follow:
``` python
def _metric_decorator(func: Callable) -> Metric:
    """Decorate a metric function from sklearn.

    Converts a metric function that uses the typical sklearn metric signature so that
    it is compatible with :py:func:`train`

    """

    def inner(y_score: np.ndarray, dmatrix: DMatrix) -> Tuple[str, float]:
        y_true = dmatrix.get_label()
        weight = dmatrix.get_weight()
        if weight.size == 0:
            return func.__name__, func(y_true, y_score)
        else:
            return func.__name__, func(y_true, y_score, sample_weight=weight)

    return inner
```
and add a test.

**Review by trivialfis — COMMENTED** [review]
Could you please help add a test in https://github.com/dmlc/xgboost/blob/master/tests/python/test_with_sklearn.py ?

**Comment by BenEfrati** [comment]
> Could you please help add a test in https://github.com/dmlc/xgboost/blob/master/tests/python/test_with_sklearn.py ?

Sure

**Review by trivialfis — APPROVED** [review]
Thank you for the nice feature!
